azure-iot-device/azure/iot/device/common/mqtt_transport.py

```diff
@@ -256,24 +266,46 @@ def reconnect(self, password=None):
         The password is not required if the transport was instantiated with an x509 certificate.
 
         :param str password: The password for reconnecting with the MQTT broker (Optional).
+
+        :raises: ConnectionFailedError if connection could not be established.
+        :raises: ConnectionDroppedError if connection is dropped during execution.
+        :raises: UnauthorizedError if there is an error authenticating.
+        :raises: ProtocolClientError if there is some other client error.
         """
         logger.info("reconnecting MQTT client")
         self._mqtt_client.username_pw_set(username=self._username, password=password)
-        rc = self._mqtt_client.reconnect()
+        try:
+            rc = self._mqtt_client.reconnect()
+        except Exception:
+            raise exceptions.ProtocolClientError("Unexpected Paho failure during reconnect")
         logger.debug("_mqtt_client.reconnect returned rc={}".format(rc))
         if rc:
+            # This could result in ConnectionFailedError, ConnectionDroppedError, UnauthorizedError
+            # or ProtocolClientError
             raise _create_error_from_rc_code(rc)
 
     def disconnect(self):
         """
         Disconnect from the MQTT broker.
+
+        :raises: ProtocolClientError if there is some client error.
         """
         logger.info("disconnecting MQTT client")
-        rc = self._mqtt_client.disconnect()
+        try:
+            rc = self._mqtt_client.disconnect()
+        except Exception:
+            raise exceptions.ProtocolClientError("Unexpected Paho failure during disconnect")
         logger.debug("_mqtt_client.disconnect returned rc={}".format(rc))
         self._mqtt_client.loop_stop()
         if rc:
-            raise _create_error_from_rc_code(rc)
```

Comment:
>_create_error_from_rc_code [](start = 18, length = 26)

I'm about to walk away so i don't have time to dig.  I want to make sure this this stack -- since this is a callback from Paho, the stack tells us volumes more than the RC does.

Reply:
Not sure what you mean here, but we do grab the text associated with the RC and use that as the message contained within the error that gets raised.

---
In reply to: [326427836](https://github.com/Azure/azure-iot-sdk-python/pull/309#discussion_r326427836) [](ancestors = 326427836)